src/zkevm_specs/evm/execution/storage.py

```diff
@@ -70,36 +70,63 @@ def sstore(instruction: Instruction):
-        else:
-            if original_value != 0:
-                if value_prev == 0:
-                    gas_refund_new = gas_refund_new - SSTORE_CLEARS_SCHEDULE
-                if value == 0:
-                    gas_refund_new = gas_refund_new + SSTORE_CLEARS_SCHEDULE
-            if original_value == value:
-                if original_value == 0:
-                    gas_refund_new = gas_refund_new + SSTORE_SET_GAS - SLOAD_GAS
-                else:
-                    gas_refund_new = gas_refund_new + SSTORE_RESET_GAS - SLOAD_GAS
+
+    # original_value, value_prev, value all are different; original_value!=0
+    nz_allne_case_refund = instruction.select(
+        instruction.is_zero(value_prev),
+        gas_refund_prev - SSTORE_CLEARS_SCHEDULE,
+        instruction.select(
+            instruction.is_zero(value),
+            gas_refund_prev + SSTORE_CLEARS_SCHEDULE,
+            gas_refund_prev,
+        ),
+    )
+    # original_value!=value_prev, value_prev!=value, original_value!=0
+    nz_ne_ne_case_refund = instruction.select(
+        1 - instruction.is_equal(original_value, value),
+        nz_allne_case_refund,
+        nz_allne_case_refund + SSTORE_RESET_GAS - SLOAD_GAS,
+    )
+    # original_value!=value_prev, value_prev!=value
+    ne_ne_case_refund = instruction.select(
+        1 - instruction.is_zero(original_value),
+        nz_ne_ne_case_refund,
+        instruction.select(
+            instruction.is_equal(original_value, value),
+            gas_refund_prev + SSTORE_SET_GAS - SLOAD_GAS,
+            gas_refund_prev,
+        ),
+    )
+    gas_refund_new = instruction.select(
+        instruction.is_equal(value_prev, value),
+        gas_refund_prev,
+        instruction.select(
+            instruction.is_equal(original_value, value_prev),
+            instruction.select(
+                (1 - instruction.is_zero(original_value)) * instruction.is_zero(value),
+                gas_refund_prev + SSTORE_CLEARS_SCHEDULE,
+                gas_refund_prev,
+            ),
+            ne_ne_case_refund,
+        ),
+    )
+
     instruction.constrain_equal(gas_refund, gas_refund_new)
 
-    if value_prev == value:
-        dynamic_gas_cost = SLOAD_GAS
-    else:
-        if original_value == value_prev:
-            if original_value == 0:
-                dynamic_gas_cost = SSTORE_SET_GAS
-            else:
-                dynamic_gas_cost = SSTORE_RESET_GAS
-        else:
-            dynamic_gas_cost = SLOAD_GAS
-    if is_warm == 0:
-        dynamic_gas_cost = dynamic_gas_cost + COLD_SLOAD_COST
+    warm_case_gas = instruction.select(
+        instruction.is_equal(value_prev, value),
+        SLOAD_GAS,
+        instruction.select(
+            instruction.is_equal(original_value, value_prev),
+            instruction.select(
+                instruction.is_zero(original_value),
+                SSTORE_SET_GAS,
+                SSTORE_RESET_GAS,
+            ),
+            SLOAD_GAS,
+        ),
+    )
```

Comment:
in fact, this part can be simplified as 
```
warm_case_gas = instruction.select(
    instruction.is_equal(value_prev, value) | (~ instruction.is_equal(original_value, value_prev)),
    SLOAD_GAS,
    instruction.select(
        instruction.is_zero(original_value),
        SSTORE_SET_GAS,
        SSTORE_RESET_GAS,
    ),
)
```

but the simplified version doesn't reflect the logic in https://eips.ethereum.org/EIPS/eip-2200 directly, so I don't know which version I should keep.

Would like to apply it if we think the simplified version is better

Reply:
Does the ` | ` in the simplified version work in the circuit?

Reply:
The logic is the same. It basically merges these two parts
1) L117~118 
2) L121~126 
together

Reply:
I see. The simplified version looks better to me. Will hold the review after you add it.